File: solutions/ruby/resistor-color-trio/1/resistor_color_trio.rb
class ResistorColorTrio
  CODING = {
    "black" => 0,
    "brown" => 1,
    "red" => 2,
    "orange" => 3,
    "yellow" => 4,
    "green" => 5,
    "blue" => 6,
    "violet" => 7,
    "grey" => 8,
    "white" => 9
  }

  SI_UNIT = {
    3 => 'kiloohms',
    6 => 'megaohms',
    9 => 'gigaohms'
  }

  attr_accessor :resistance, :divisor, :si_unit
  
  def initialize(colors)
    unit_place = 10 ** CODING[colors[2]]
    
    @resistance = (
      CODING[colors[0]].to_s + CODING[colors[1]].to_s
    ).to_i * unit_place

    @divisor = get_divisor(CODING[colors[2]])   
    @si_unit = get_si_units(CODING[colors[2]])    
  end

  def label
    "Resistor value: #{resistance/divisor} #{si_unit}"
  end

  private
  def get_si_units(third)
    return "gigaohms" if third == 9
    return "megaohms" if third >= 6 && third < 9
    return "kiloohms" if third >= 2 && third < 6
    return "ohms" if third >= 0 && third < 2
  end    

  def get_divisor(third)
    return 10 ** 9 if third == 9
    return 10 ** 6 if third >= 6 && third < 9
    return 10 ** 3 if third >= 2 && third < 6
    return 1 if third >= 0 && third < 2
  end
end  

File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
export const CODING: { [key: string]: number } = {
  black: 0, brown: 1, red: 2,
  orange: 3, yellow: 4, green: 5,
  blue: 6, violet: 7, grey: 8, white: 9
};

export const UNITS = [
  { threshold: 9, name: "gigaohms", divisor: 1e9 },
  { threshold: 6, name: "megaohms", divisor: 1e6 },
  { threshold: 2, name: "kiloohms", divisor: 1e3 },
  { threshold: 0, name: "ohms", divisor: 1 }
];

export interface ResistorValue {
  resistance: number;
  unitName: string;
  divisor: number;
  label: () => string;
}

export function decodedResistorValue(colors: string[]): string {

  const baseValue = Number(`${CODING[colors[0]]}${CODING[colors[1]]}`)
  const multiplierExp = CODING[colors[2]];

  const resistance = baseValue * Math.pow(10, multiplierExp);
  const unit = UNITS.find(u => multiplierExp >= u.threshold)!;

  return `${resistance / unit.divisor} ${unit.name}`
}